ccaas_builder/cmd/build/main.go

```diff
@@ -0,0 +1,189 @@
+/*
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+	"text/template"
+
+	"github.com/otiai10/copy"
+)
+
+var logger = log.New(os.Stderr, "", 0)
+
+type ChaincodeMetadata struct {
+	Type string `json:"type"`
+}
+
+// Connection structure is used to represent the
+// connection.json file that is supplied in the
+// chaincode package.
+type Connection struct {
+	Address     string `json:"address"`
+	DialTimeout string `json:"dial_timeout"`
+	TLS         bool   `json:"tls_required"`
+	ClientAuth  bool   `json:"client_auth_required"`
+	RootCert    string `json:"root_cert"`
+	ClientKey   string `json:"client_key"`
+	ClientCert  string `json:"client_cert"`
+}
+
+type Config struct {
+	PeerName string
+}
+
+func main() {
+	logger.Println("::Build")
+
+	if err := run(); err != nil {
+		logger.Printf("::Error: %v\n", err)
+		os.Exit(1)
+	}
+
+	logger.Printf("::Build phase completed")
+
+}
+
+func run() error {
+	if len(os.Args) < 4 {
+		return fmt.Errorf("incorrect number of arguments")
+	}
+
+	sourceDir, metadataDir, outputDir := os.Args[1], os.Args[2], os.Args[3]
+
+	connectionSrcFile := filepath.Join(sourceDir, "/connection.json")
+	metadataFile := filepath.Clean(filepath.Join(metadataDir, "metadata.json"))
+	connectionDestFile := filepath.Join(outputDir, "/connection.json")
+	metainfoSrcDir := filepath.Join(sourceDir, "META-INF")
+	metainfoDestDir := filepath.Join(outputDir, "META-INF")
+
+	// Process and check the metadata file, then copy to the output location
+	metadataFileContents, err := ioutil.ReadFile(metadataFile)
+	if err != nil {
+		return err
+	}
+
+	var metadata ChaincodeMetadata
+	if err := json.Unmarshal(metadataFileContents, &metadata); err != nil {
+		return err
+	}
+
+	if strings.ToLower(metadata.Type) != "ccaas" {
+		return fmt.Errorf("chaincode type should be ccaas, it is %s", metadata.Type)
+	}
+
+	if err := copy.Copy(metadataDir, outputDir); err != nil {
+		return fmt.Errorf("failed to copy build metadata folder: %s", err)
+	}
+
+	if _, err := os.Stat(metainfoSrcDir); !os.IsNotExist(err) {
+		if err := copy.Copy(metainfoSrcDir, metainfoDestDir); err != nil {
+			return fmt.Errorf("failed to copy build META-INF folder: %s", err)
+		}
+	}
+
+	// Process and update the connections file
+	fileInfo, err := os.Stat(connectionSrcFile)
+	if err != nil {
+		return fmt.Errorf("connection.json not found in source folder: %s", err)
```

Comment:
Good, this is what I mentioned should be done in detect, but using errors.WithMessage() to augment the error message.